Automattic/woocommerce-payments
# Add correct metadata to orders for Sofort transactions

Fixes #1878 

#### Changes proposed in this Pull Request


***Fixing Order Metadata for Sofort Transactions***


This PR mirrors the changes suggested in #1859 to allow payments made using the Sofort payment method to appear correctly in the WCPay Transactions list. This is achieved essentially by adding the `charge_id` for the transaction, among other fields, to the `WC_Order` upon completing a checkout. 





#### Testing instructions





1. Please pull the changes from this PR and run them in parallel to the latest version of the WCPay Payments Server. To enable the Sofort payment gateway, you need to set the Sofort feature flag by adding a row to your site's `wp_options` table with an `option_name` of `_wcpay_feature_sofort` and an  `option_value` of `1`. You will also need to set your store's currency to use Euros.

![Sofort feature flag row](https://cdn-std.droplr.net/files/acc_1104206/qUa8co)
_Row in `wp_options` table to set Sofort feature flag_

2. Purchase an item from your store using Sofort as a payment method. Successfully authorise the transaction once redirected to Stripe to complete the payment. If you view the order details for that corresponding order created, its status should immediately be `On Hold`.
3. In test mode, [Sofort Payment Intents remain in `processing` state for three minutes, until this state changes to `succeeded`](https://stripe.com/docs/payments/sofort/accept-a-payment?platform=web#fulfillment). Once this state has changed, this fully completes the payment and you should see the Order Status update to `Processing`. You should be able to see this reflected in the order notes.

![Order notes update as payment status changes](https://cdn-std.droplr.net/files/acc_1104206/fMgqtP)
_Order notes reflects order status changes as payment transaction progresses_

4. Now go to the list of Transactions at **Payments > Transactions** and you should see an entry matching your completed order, with the correct order ID, customer name, price (in converted currency), and Sofort logo.

![Sofort transaction list detail](https://cdn-std.droplr.net/files/acc_1104206/G6dMZB)
_Sofort transaction summary appears in list_

5. You can click into the transaction and confirm the details match the purchase, but otherwise that's it. Maybe take a break and treat yourself to a nice snack now, if you feel like it. You deserve it.

-------------------
